### ISSUE TITLE
RHEL6 - API - unreliable responding

### ISSUE THREAD

**Issue by distahl** [Icinga/icinga2#8247]
## Describe the bug

As far as I know, you don't officially support EL6 platforms anymore, but as you still provide rpms for them, I thought I will try to report this anyway.

On RHEL6 I get this weired issue if I try to query icinga's api. With e.g.
`curl -k -u 'user:pass' -H 'Accept: application/json' -H 'X-HTTP-Method-Override: GET'  -X POST  'https://127.0.0.1:5665/v1/objects/Hosts' -d '{ "attrs": [ "name" ], "pretty": true}'`

Sometimes the request works fine and sometimes I get the following error:
`curl: (52) NSS: client certificate not found (nickname not specified)`

I also compiled myself a newer curl with openssl instead of nss to test and this leads to a similar issue with different error message:
`curl: (52) Empty reply from server`

Btw.: 
As Icingaweb2 is using also php's curl implementation (afaik) for api calls. 
You can also see this error in the icingaweb2 logs.

## To Reproduce

Repeatly call `curl -k -u 'user:pass' -H 'Accept: application/json' -H 'X-HTTP-Method-Override: GET'  -X POST  'https://127.0.0.1:5665/v1/objects/Hosts' -d '{ "attrs": [ "name" ], "pretty": true}'`. The first call most likely will work, but the following calls will often fail. If you wait some time there is a higher chance, that the next call will work again.

## Expected behavior

Reliably give output on every call.

## Your Environment

```
icinga2 - The Icinga 2 network monitoring daemon (version: 2.12.0-1)

Copyright (c) 2012-2020 Icinga GmbH (https://icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

System information:
  Platform: Red Hat Enterprise Linux Server
  Platform version: 6.10 (Santiago)
  Kernel: Linux
  Kernel version: 2.6.32-754.29.2.el6.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 7.3.1
  Build host: runner-hh8q3bz2-project-322-concurrent-0
  OpenSSL version: OpenSSL 1.0.1e-fips 11 Feb 2013
```
* Enabled features (`icinga2 feature list`):
`Disabled features: compatlog debuglog elasticsearch gelf graphite icingadb livestatus opentsdb statusdata syslog`
`Enabled features: api checker command ido-mysql influxdb mainlog notification perfdata`

## Additional context

**NSS Curl:**
```
curl -V
curl 7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.44 zlib/1.2.3 libidn/1.18 libssh2/1.4.2
Protocols: tftp ftp telnet dict ldap ldaps http file https ftps scp sftp 
Features: GSS-Negotiate IDN IPv6 Largefile NTLM SSL libz
```

**Openssl Curl:**
```
curl -V
curl 7.29.0 (x86_64-redhat-linux-gnu) libcurl/7.29.0 OpenSSL/1.0.2k zlib/1.2.3 libidn/1.18 libssh2/1.4.2
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smtp smtps telnet tftp 
Features: AsynchDNS GSS-Negotiate IDN IPv6 Largefile NTLM NTLM_WB SSL libz unix-sockets
```

**Comment by Al2Klimov** [comment]
Hello @distahl and thank you for reporting!

How many hosts do you have?

Best,
AK

**Comment by distahl** [comment]
Not very much. This is only a testing environment:

```
icinga2 daemon --validate
[2020-09-21 14:29:57 +0200] information/cli: Icinga application loader (version: 2.12.0-1)
[2020-09-21 14:29:57 +0200] information/cli: Loading configuration file(s).
[2020-09-21 14:29:57 +0200] information/ConfigItem: Committing config item(s).
[2020-09-21 14:29:57 +0200] information/ApiListener: My API identity: xyz
[...]
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 NotificationComponent.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 739 Hosts.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 Downtime.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 6 NotificationCommands.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 FileLogger.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 Comment.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 955 Notifications.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 IcingaApplication.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1585 HostGroups.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 PerfdataWriter.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 5390 Dependencies.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 CheckerComponent.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 4 Zones.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 2 Endpoints.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 ExternalCommandListener.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 IdoMysqlConnection.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 3 ApiUsers.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 ApiListener.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 332 CheckCommands.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 InfluxdbWriter.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 3 TimePeriods.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 5 UserGroups.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 2 Users.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 4244 Services.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 3 ServiceGroups.
[2020-09-21 14:30:01 +0200] information/ConfigItem: Instantiated 1 ScheduledDowntime.
[2020-09-21 14:30:01 +0200] information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
[2020-09-21 14:30:01 +0200] information/cli: Finished validating the configuration file(s)
```

**Comment by Al2Klimov** [comment]
> This is only a testing environment:

Is there anything confidential? Could you tar/gz `/etc/icinga2` and `/var/*/icinga2` and upload them for us?

**Comment by distahl** [comment]
> Is there anything confidential? Could you tar/gz `/etc/icinga2` and `/var/*/icinga2` and upload them for us?
Unfortunately, yes, there is confidential data among the configs. So I can't upload this. Sorry.

**Comment by Al2Klimov** [comment]
Notes:

* yum update doesn't help
* v2.11 downgrade doesn't help

**Comment by Al2Klimov** [comment]
> I also compiled myself a newer curl with openssl instead of nss

This makes it working for me.

> `curl: (52) Empty reply from server`

Please could you enable the debug log and share the entries around this error?

> As Icingaweb2 is using also php's curl implementation (afaik) for api calls.
> You can also see this error in the icingaweb2 logs.

[CentOS 6 is not supported as master.](https://icinga.com/subscription/support-details/)

**Comment by distahl** [comment]
> > I also compiled myself a newer curl with openssl instead of nss
>
> This makes it working for me.

Hmm...Interesting. I used the source rpm of the centos 7 and changed some parts to make it compile (rpmbuild) with openssl. And also did the same with openssl, which is used by curl. Maybe this is a difference. 
If you are interested I can try to upload those rpms for you.

> Please could you enable the debug log and share the entries around this error?

Do you mean verbose mode of curl or debug log of icinga?

If you mean icinga. I did this some days ago and wasn't able to find the right lines. Icinga is spamming the file so much with debugging info, which makes it nearly impossible for me to find the right line without knowing what to search for. Do you maybe have some hint for what I should grep?

If you are interested in verbose logging of curl, here you go:

Curl with NSS - working call:
```
* About to connect() to 127.0.0.1 port 5665 (#0)
*   Trying 127.0.0.1... connected
* Connected to 127.0.0.1 (127.0.0.1) port 5665 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* warning: ignoring value of ssl.verifyhost
* skipping SSL peer certificate verification
* NSS: client certificate not found (nickname not specified)
* SSL connection using TLS_RSA_WITH_AES_256_GCM_SHA384
* Server certificate:
*       subject: CN=xyz
*       start date: Jan 04 09:26:56 2017 GMT
*       expire date: Jan 01 09:26:56 2032 GMT
*       common name: xyz
*       issuer: CN=Icinga CA
* Server auth using Basic with user 'SomeUser'
> POST /v1/objects/Hosts HTTP/1.1
> Authorization: Basic xxxx
> User-Agent: curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.44 zlib/1.2.3 libidn/1.18 libssh2/1.4.2
> Host: 127.0.0.1:5665
> Accept: application/json
> X-HTTP-Method-Override: GET
> Content-Length: 38
> Content-Type: application/x-www-form-urlencoded
> 
} [data not shown]
< HTTP/1.1 200 OK
< Server: Icinga/2.12.0-1
< Content-Type: application/json
< Content-Length: 158753
< 
{ [data not shown]

[... api json result here ...]

* Connection #0 to host 127.0.0.1 left intact
* Closing connection #0
```

Curl with NSS - failed call:
```
* About to connect() to 127.0.0.1 port 5665 (#0)
*   Trying 127.0.0.1... connected
* Connected to 127.0.0.1 (127.0.0.1) port 5665 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* warning: ignoring value of ssl.verifyhost
* skipping SSL peer certificate verification
* NSS: client certificate not found (nickname not specified)
* SSL connection using TLS_RSA_WITH_AES_256_GCM_SHA384
* Server certificate:
*       subject: CN=xyz
*       start date: Jan 04 09:26:56 2017 GMT
*       expire date: Jan 01 09:26:56 2032 GMT
*       common name: xyz
*       issuer: CN=Icinga CA
* Server auth using Basic with user 'SomeUser'
> POST /v1/objects/Hosts HTTP/1.1
> Authorization: Basic xxxx
> User-Agent: curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.44 zlib/1.2.3 libidn/1.18 libssh2/1.4.2
> Host: 127.0.0.1:5665
> Accept: application/json
> X-HTTP-Method-Override: GET
> Content-Length: 38
> Content-Type: application/x-www-form-urlencoded
> 
} [data not shown]
* Empty reply from server
* Connection #0 to host 127.0.0.1 left intact

curl: (52) NSS: client certificate not found (nickname not specified)
* Closing connection #0
```

Curl with openssl - working call:
```
* About to connect() to 127.0.0.1 port 5665 (#0)
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 5665 (#0)
* successfully set certificate verify locations:
*   CAfile: /etc/opt/rh/rh-curl/pki/tls/certs/ca-bundle.crt
  CApath: none
* SSLv3, TLS Unknown, Unknown (22):
} [data not shown]
* SSLv3, TLS handshake, Client hello (1):
} [data not shown]
* SSLv2, Unknown (22):
{ [data not shown]
* SSLv3, TLS handshake, Server hello (2):
{ [data not shown]
* SSLv2, Unknown (22):
{ [data not shown]
* SSLv3, TLS handshake, CERT (11):
{ [data not shown]
* SSLv2, Unknown (22):
{ [data not shown]
* SSLv3, TLS handshake, Request CERT (13):
{ [data not shown]
* SSLv3, TLS handshake, Server finished (14):
{ [data not shown]
* SSLv2, Unknown (22):
} [data not shown]
* SSLv3, TLS handshake, CERT (11):
} [data not shown]
* SSLv2, Unknown (22):
} [data not shown]
* SSLv3, TLS handshake, Client key exchange (16):
} [data not shown]
* SSLv2, Unknown (20):
} [data not shown]
* SSLv3, TLS change cipher, Client hello (1):
} [data not shown]
* SSLv2, Unknown (22):
} [data not shown]
* SSLv3, TLS handshake, Finished (20):
} [data not shown]
* SSLv2, Unknown (20):
{ [data not shown]
* SSLv3, TLS change cipher, Client hello (1):
{ [data not shown]
* SSLv2, Unknown (22):
{ [data not shown]
* SSLv3, TLS handshake, Finished (20):
{ [data not shown]
* SSL connection using AES256-GCM-SHA384
* Server certificate:
*        subject: CN=xyz
*        start date: 2017-01-04 09:26:56 GMT
*        expire date: 2032-01-01 09:26:56 GMT
*        issuer: CN=Icinga CA
*        SSL certificate verify result: self signed certificate in certificate chain (19), continuing anyway.
* Server auth using Basic with user 'SomeUser'
* SSLv2, Unknown (23):
} [data not shown]
> POST /v1/objects/Hosts HTTP/1.1
> Authorization: Basic xxxx
> User-Agent: curl/7.29.0
> Host: 127.0.0.1:5665
> Accept: application/json
> X-HTTP-Method-Override: GET
> Content-Length: 38
> Content-Type: application/x-www-form-urlencoded
> 
} [data not shown]
* upload completely sent off: 38 out of 38 bytes
* SSLv2, Unknown (23):
{ [data not shown]
< HTTP/1.1 200 OK
< Server: Icinga/2.12.0-1
< Content-Type: application/json
< Content-Length: 158753
< 
{ [data not shown]
* SSLv2, Unknown (23):
{ [data not shown]
* SSLv2, Unknown (23):
{ [data not shown]
[... many more of the two lines above ...]

[... api result json here ...]

* Connection #0 to host 127.0.0.1 left intact
```

Curl with openssl - failed call:
```
* About to connect() to 127.0.0.1 port 5665 (#0)
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 5665 (#0)
* successfully set certificate verify locations:
*   CAfile: /etc/opt/rh/rh-curl/pki/tls/certs/ca-bundle.crt
  CApath: none
* SSLv3, TLS Unknown, Unknown (22):
* SSLv3, TLS handshake, Client hello (1):
* SSLv2, Unknown (22):
* SSLv3, TLS handshake, Server hello (2):
* SSLv2, Unknown (22):
* SSLv3, TLS handshake, CERT (11):
* SSLv2, Unknown (22):
* SSLv3, TLS handshake, Request CERT (13):
* SSLv3, TLS handshake, Server finished (14):
* SSLv2, Unknown (22):
* SSLv3, TLS handshake, CERT (11):
* SSLv2, Unknown (22):
* SSLv3, TLS handshake, Client key exchange (16):
* SSLv2, Unknown (20):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv2, Unknown (22):
* SSLv3, TLS handshake, Finished (20):
* SSLv2, Unknown (20):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv2, Unknown (22):
* SSLv3, TLS handshake, Finished (20):
* SSL connection using AES256-GCM-SHA384
* Server certificate:
*        subject: CN=xyz
*        start date: 2017-01-04 09:26:56 GMT
*        expire date: 2032-01-01 09:26:56 GMT
*        issuer: CN=Icinga CA
*        SSL certificate verify result: self signed certificate in certificate chain (19), continuing anyway.
* Server auth using Basic with user 'SomeUser'
* SSLv2, Unknown (23):
> POST /v1/objects/Hosts HTTP/1.1
> Authorization: Basic xxxxxx
> User-Agent: curl/7.29.0
> Host: 127.0.0.1:5665
> Accept: application/json
> X-HTTP-Method-Override: GET
> Content-Length: 38
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 38 out of 38 bytes
* SSLv2, Unknown (21):
* SSLv3, TLS alert, Client hello (1):
* Empty reply from server
* Connection #0 to host 127.0.0.1 left intact
curl: (52) Empty reply from server
```

**Comment by Al2Klimov** [comment]
> Icinga is spamming the file so much with debugging info, which makes it nearly impossible for me to find the right line without knowing what to search for.

Could you start a `tail -f` on the main log (icinga2.log), provoke a curl failure, stop the tail and share it?

**Comment by distahl** [comment]
> Could you start a `tail -f` on the main log (icinga2.log), provoke a curl failure, stop the tail and share it?

I just tried, but there is nothing if the curl call fails. Only if the call succeeds I can see a logline with a request.

**Comment by Al2Klimov** [comment]
Nothing? Absolutely nothing? Not even like "new connection"?

**Comment by distahl** [comment]
Yes, really nothing.

**Comment by distahl** [comment]
I've made some more tests:

I downgraded icinga2 to v2.11.5, 2.11.4, 2.11.3, 2.11.2, 2.11.1, 2.11.0 and always tested if curl connection is reliable again. But to my surprise on all versions I have the same issue. So it is not an issue which only applies to v2.12.0.

Another test I've done was doing a curl api call, from rhel 6 system, to another icinga system, which is running v2.12.0 on centos7.  And this was reliably working. So curl on rhel6 isn't the problem. It seems to be a combination of icinga and the OS on rhel6, which is causing this. But cannot see any error messages which could leed to the root cause.
